modules/fedimint-lnv2-client/src/lib.rs

```diff
@@ -433,7 +437,36 @@ fn generate_ephemeral_tweak(static_pk: PublicKey) -> ([u8; 32], PublicKey) {
 }
 
 impl LightningClientModule {
-    async fn select_gateway(&self) -> Result<(SafeUrl, RoutingInfo), SelectGatewayError> {
+    /// This method updates the mapping from lightning node public keys to
+    /// gateway api endpoints maintained in the module database once a day. When
+    /// paying an invoice this enables the client to select the gateway that has
+    /// created the invoice, if possible, such that the payment does not go
+    /// over lightning, reducing fees and latency.
+    ///
+    /// Client integrators are expected to call this function in a spawned task.
+    pub async fn update_gateway_map(&self) -> ! {
+        loop {
+            if let Ok(gateways) = self.module_api.gateways().await {
+                let mut dbtx = self.client_ctx.module_db().begin_transaction().await;
+
+                for gateway in gateways {
+                    if let Ok(Some(routing_info)) = self.routing_info(&gateway).await {
+                        dbtx.insert_entry(&GatewayKey(routing_info.lightning_public_key), &gateway)
+                            .await;
+                    }
+                }
```

Comment:
This code never invalidates gateways that got removed/are unreachable or am I missing something? That doesn't seem optimal.

I'd probably try to reach all of them first, then if any of them was reachable remove all DB entries and insert new ones. If none was reachable assume we are just went offline right after fetching the gateways, don't change the DB and wait for the next time the update is run.

Reply:
When the gateway is selected to pay an invoice it will only consider the gateways returned by the federation at that moment, so an outdated entry will be ignored.

Reply:
I agree, sorry about that. Should have opened as draft.